minetest/minetest
# Use scoped app storage on Android

**This must be merged and released with sufficient time before November 2021, to avoid data loss**

From November 2021, the Play Store will [no longer be accepting](https://developer.android.com/distribute/play-policies#APILevel30) updates to apps which use the deprecated getExternalStorageDirectory() API.

Therefore, this commit replaces uses of the deprecated API with the new scoped API (`getExternalFilesDir()` and `getExternalCacheDir()`). It also provides a temporary migration to move user data from the shared external directory to new storage.

See https://developer.android.com/training/data-storage/ and https://developer.android.com/about/versions/11/privacy/storage

Fixes #2097, fixes #11417, fixes #11299. Potentially fixes #11462

## To do

- [x] Improve progress indication when migrating. Currently, nothing appears on screen. There should either be a progress bar or an indeterminate spinner on screen whilst the unzip service is doing something
- [x] Check externalFilesDir state before launching app, to gracefully show an error
- [x] Allow the UI to exit and recreate without requeing the service
- [ ] Check fault tolerance

## How to test

:warning: Warning! Make sure to backup any important files before testing :warning:
:warning: This will move any worlds from the shared storage to the app storage. When the app is uninstalled, all stored data will be deleted - including worlds. Additionally, I can't guarantee it won't mess up :warning:

APK builds: https://github.com/rubenwardy/minetest/releases/tag/5.5.0-scoped1

Perform the following on Android 11, 10, and 4/5:

* Check migration
  * Create worlds on the old version of the app
  * Upgrade
  * Run Minetest again, check that the worlds still appear
* Check new install
  * With no Minetest data, check that you can install the PR version and it works
